### ISSUE TITLE
(EDA → Modelling → Evaluasi)

### ISSUE THREAD

**Issue by taufik054-jpg** [alan-turing-institute/rds-course#173]
"""
turnover_pipeline.py
Kode lengkap: EDA -> Imputasi -> Feature Engineering -> Modelling -> Evaluasi
Output disimpan di /mnt/data/
"""

import os
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

# plotting
import matplotlib.pyplot as plt
import seaborn as sns

# stats
from scipy import stats

# preprocessing & imputasi
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.experimental import enable_iterative_imputer  # noqa
from sklearn.impute import IterativeImputer

# models & metrics
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (accuracy_score, precision_score, recall_score, f1_score,
                             roc_auc_score, confusion_matrix, classification_report,
                             roc_curve, auc)
import joblib

# imbalance
try:
    from imblearn.over_sampling import SMOTE
    from imblearn.pipeline import Pipeline as ImbPipeline
    IMBLEARN_AVAILABLE = True
except Exception:
    IMBLEARN_AVAILABLE = False

# xgboost optional
try:
    import xgboost as xgb
    XGBOOST_AVAILABLE = True
except Exception:
    XGBOOST_AVAILABLE = False

# Paths
OUTPUT_DIR = "/mnt/data"
os.makedirs(OUTPUT_DIR, exist_ok=True)

# Attempt to find dataset in /mnt/data
def find_dataset():
    candidates = []
    for fname in os.listdir("/mnt/data"):
        if fname.lower().endswith((".csv", ".xlsx", ".xls")):
            candidates.append(os.path.join("/mnt/data", fname))
    return candidates

def load_dataset():
    files = find_dataset()
    if files:
        path = files[0]
        print(f"Found dataset: {path} — attempting to load.")
        if path.lower().endswith(".csv"):
            return pd.read_csv(path)
        else:
            return pd.read_excel(path)
    else:
        print("No dataset found in /mnt/data. Generating synthetic dataset.")
        return generate_synthetic_data(seed=42, n=1200)

def generate_synthetic_data(seed=42, n=1000):
    np.random.seed(seed)
    # generate age (21-55), tenure (0.5-12), salary (in million), distance (km)
    age = np.random.normal(loc=34, scale=6, size=n).astype(int)
    age = np.clip(age, 21, 60)
    tenure = np.abs(np.random.normal(loc=4, scale=2, size=n))
    tenure = np.clip(tenure, 0.2, 20).round(2)
    # salary in juta rupiah
    base = np.random.normal(loc=7.5, scale=2.2, size=n)
    # create skew by adding log-normal tail
    salary = np.clip(base + np.random.lognormal(mean=0.1, sigma=0.6, size=n), 3, 25).round(2)
    distance = np.abs(np.random.exponential(scale=8, size=n)).round(2)
    distance = np.clip(distance, 0.5, 60)

    # create turnover probability influenced by low salary, long distance, low tenure
    prob = (0.25 * (distance / (distance.max()+1))
            + 0.4 * (1 - (salary - salary.min())/(salary.max()-salary.min()))
            + 0.35 * (1 - np.tanh(tenure/5)))
    prob = (prob - prob.min()) / (prob.max() - prob.min())  # normalize 0-1

    turnover = np.random.binomial(1, prob)

    df = pd.DataFrame({
        "age": age,
        "tenure": tenure,
        "salary_million": salary,
        "distance_km": distance,
        "turnover": turnover
    })

    # introduce some missing values randomly
    for col in ["tenure", "salary_million", "distance_km"]:
        df.loc[df.sample(frac=0.06, random_state=seed).index, col] = np.nan

    # add a categorical column department for illustration
    depts = ["HR", "Engineering", "Sales", "Finance", "Operations"]
    df["department"] = np.random.choice(depts, size=n, p=[0.12,0.35,0.25,0.13,0.15])

    return df

# ----------------------------------------------
# EDA functions
# ----------------------------------------------
def save_histograms(df, numeric_cols, outdir=OUTPUT_DIR):
    for col in numeric_cols:
        plt.figure(figsize=(6,4))
        sns.histplot(df[col].dropna(), kde=True)
        plt.title(f"Histogram: {col}")
        plt.xlabel(col)
        plt.tight_layout()
        p = os.path.join(outdir, f"hist_{col}.png")
        plt.savefig(p)
        plt.close()

def save_boxplots(df, numeric_cols, outdir=OUTPUT_DIR):
    for col in numeric_cols:
        plt.figure(figsize=(6,3))
        sns.boxplot(x=df[col])
        plt.title(f"Boxplot: {col}")
        plt.tight_layout()
        p = os.path.join(outdir, f"box_{col}.png")
        plt.savefig(p)
        plt.close()

def save_correlation_heatmap(df, numeric_cols, outdir=OUTPUT_DIR):
    corr = df[numeric_cols].corr()
    plt.figure(figsize=(6,5))
    sns.heatmap(corr, annot=True, fmt=".2f", cmap="coolwarm")
    plt.title("Correlation Heatmap")
    plt.tight_layout()
    p = os.path.join(outdir, "corr_heatmap.png")
    plt.savefig(p)
    plt.close()

def compute_skewness(df, numeric_cols):
    skew_dict = {}
    for c in numeric_cols:
        skew_dict[c] = float(df[c].dropna().skew())
    return skew_dict

# ----------------------------------------------
# Main pipeline
# ----------------------------------------------
def main():
    df = load_dataset()
    print("Dataset shape:", df.shape)
    print(df.head())

    # Ensure expected columns exist; if not, try to adapt
    expected_cols = ["age", "tenure", "salary_million", "distance_km", "turnover"]
    missing_expected = [c for c in expected_cols if c not in df.columns]
    if missing_expected:
        print("Warning: dataset missing expected columns:", missing_expected)
        # try to map some common alternate names
        mapping = {}
        if "gaji" in df.columns:
            mapping["gaji"] = "salary_million"
        if "lama_bekerja" in df.columns:
            mapping["lama_bekerja"] = "tenure"
        if mapping:
            df = df.rename(columns=mapping)
            print("Renamed columns according to mapping:", mapping)

    # Basic EDA
    numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    print("\nNumeric columns detected:", numeric_cols)

    desc = df.describe(include='all')
    desc.to_csv(os.path.join(OUTPUT_DIR, "describe.csv"))
    print("Saved describe.csv")

    # Save histograms and boxplots
    num_cols_for_plots = [c for c in numeric_cols if c != "turnover"]
    save_histograms(df, num_cols_for_plots)
    save_boxplots(df, num_cols_for_plots)
    save_correlation_heatmap(df, [c for c in numeric_cols if c != "turnover"])

    # Compute skewness
    skew = compute_skewness(df, num_cols_for_plots)
    print("\nSkewness per numeric column:")
    for k,v in skew.items():
        print(f"  {k}: {v:.3f}")

    # Correlation with target
    if "turnover" in df.columns:
        corr_with_target = {}
        for c in num_cols_for_plots:
            corr_with_target[c] = df[[c, "turnover"]].dropna().corr().iloc[0,1]
        print("\nCorrelation with turnover:")
        for k,v in corr_with_target.items():
            print(f"  {k}: {v:.3f}")

    # -------------------------
    # Missing value handling
    # -------------------------
    print("\nMissing values per column:")
    print(df.isna().sum())

    # Choose imputasi strategy:
    # Options: "median", "knn", "iterative"
    IMPUTE_STRATEGY = "iterative"  # change to "median" or "knn" if desired

    df_imputed = df.copy()
    numeric_features = ["age", "tenure", "salary_million", "distance_km"]
    cat_features = [c for c in df.columns if df[c].dtype == 'object' or df[c].dtype.name == 'category']

    if IMPUTE_STRATEGY == "median":
        print("Using median imputation for numeric features.")
        med = SimpleImputer(strategy="median")
        df_imputed[numeric_features] = med.fit_transform(df_imputed[numeric_features])
    elif IMPUTE_STRATEGY == "knn":
        print("Using KNN imputation for numeric features (n_neighbors=5).")
        knn = KNNImputer(n_neighbors=5)
        df_imputed[numeric_features] = knn.fit_transform(df_imputed[numeric_features])
    else:
        print("Using Iterative (Predictive) Imputation for numeric features.")
        it = IterativeImputer(random_state=0, max_iter=10)
        df_imputed[numeric_features] = it.fit_transform(df_imputed[numeric_features])

    # confirm no missing in numeric now
    print("Missing after imputation:")
    print(df_imputed.isna().sum())

    # -------------------------
    # Feature Engineering
    # -------------------------
    # 1) salary category
    bins = [0, 5, 10, 1000]  # salary in million
    labels = ["Low", "Medium", "High"]
    df_imputed["salary_cat"] = pd.cut(df_imputed["salary_million"], bins=bins, labels=labels, include_lowest=True)

    # 2) log transforms for skewed features
    # add tiny constant to avoid log(0)
    df_imputed["log_salary"] = np.log1p(df_imputed["salary_million"])
    df_imputed["log_distance"] = np.log1p(df_imputed["distance_km"])

    # 3) interaction features
    df_imputed["age_tenure_interaction"] = df_imputed["age"] * df_imputed["tenure"]
    df_imputed["distance_per_tenure"] = df_imputed["distance_km"] / (df_imputed["tenure"] + 0.1)

    # 4) ensure target is integer
    df_imputed["turnover"] = df_imputed["turnover"].astype(int)

    # -------------------------
    # Prepare X, y
    # -------------------------
    target = "turnover"
    drop_cols = []  # columns to drop if any
    X = df_imputed.drop(columns=[target] + drop_cols)
    y = df_imputed[target]

    # split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

    print("\nTrain class distribution:")
    print(y_train.value_counts(normalize=True))
    print("\nTest class distribution:")
    print(y_test.value_counts(normalize=True))

    # -------------------------
    # Preprocessing pipeline
    # -------------------------
    numeric_transformer = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="median")),  # numeric already imputed, but keep for safety
        ("scaler", StandardScaler())
    ])

    categorical_transformer = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("onehot", OneHotEncoder(handle_unknown="ignore"))
    ])

    # choose columns for pipeline (some engineered features are numeric)
    preproc = ColumnTransformer(transformers=[
        ("num", numeric_transformer, ["age", "tenure", "salary_million", "distance_km",
                                     "log_salary", "log_distance", "age_tenure_interaction", "distance_per_tenure"]),
        ("cat", categorical_transformer, ["salary_cat", "department"])
    ], remainder="drop")

    # -------------------------
    # Optionally handle imbalance with SMOTE
    # -------------------------
    USE_SMOTE = True if IMBLEARN_AVAILABLE else False
    if USE_SMOTE:
        print("imblearn detected — will use SMOTE in pipeline.")
    else:
        print("imblearn not available — proceeding without SMOTE. To enable SMOTE, install imblearn.")

    # -------------------------
    # Models and training
    # -------------------------
    # 1) Logistic Regression
    pipe_log = Pipeline(steps=[("preproc", preproc),
                               ("clf", LogisticRegression(max_iter=1000, class_weight="balanced"))])

    pipe_log.fit(X_train, y_train)
    y_pred_log = pipe_log.predict(X_test)
    y_proba_log = pipe_log.predict_proba(X_test)[:,1]

    # 2) Random Forest with simple GridSearch
    rf = RandomForestClassifier(random_state=42, n_jobs=-1, class_weight=None)
    pipe_rf = Pipeline(steps=[("preproc", preproc), ("clf", rf)])

    # simple grid
    param_grid = {
        "clf__n_estimators": [100, 200],
        "clf__max_depth": [5, 10, None],
        "clf__min_samples_split": [2, 5]
    }
    grid = GridSearchCV(pipe_rf, param_grid, cv=3, scoring="roc_auc", n_jobs=-1)
    grid.fit(X_train, y_train)
    best_rf = grid.best_estimator_
    print("Best RF params:", grid.best_params_)

    y_pred_rf = best_rf.predict(X_test)
    y_proba_rf = best_rf.predict_proba(X_test)[:,1]

    # 3) XGBoost (optional)
    if XGBOOST_AVAILABLE:
        print("XGBoost available — training XGBClassifier.")
        xgb_clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric="logloss", random_state=42)
        pipe_xgb = Pipeline(steps=[("preproc", preproc), ("clf", xgb_clf)])
        pipe_xgb.fit(X_train, y_train)
        y_pred_xgb = pipe_xgb.predict(X_test)
        y_proba_xgb = pipe_xgb.predict_proba(X_test)[:,1]
    else:
        print("XGBoost not available — skipping.")

    # -------------------------
    # Evaluation helper
    # -------------------------
    def eval_and_save(y_true, y_pred, y_proba, prefix):
        acc = accuracy_score(y_true, y_pred)
        prec = precision_score(y_true, y_pred, zero_division=0)
        rec = recall_score(y_true, y_pred, zero_division=0)
        f1 = f1_score(y_true, y_pred, zero_division=0)
        roc = roc_auc_score(y_true, y_proba) if len(np.unique(y_true)) > 1 else np.nan

        print(f"\n-- {prefix} Evaluation --")
        print(f"Accuracy: {acc:.4f}")
        print(f"Precision: {prec:.4f}")
        print(f"Recall: {rec:.4f}")
        print(f"F1-score: {f1:.4f}")
        print(f"ROC-AUC: {roc:.4f}")
        print("Confusion Matrix:")
        cm = confusion_matrix(y_true, y_pred)
        print(cm)
        print("Classification Report:")
        print(classification_report(y_true, y_pred, zero_division=0))

        # Save metrics to csv
        metrics_df = pd.DataFrame([{
            "model": prefix,
            "accuracy": acc,
            "precision": prec,
            "recall": rec,
            "f1": f1,
            "roc_auc": roc
        }])
        metrics_df.to_csv(os.path.join(OUTPUT_DIR, f"metrics_{prefix}.csv"), index=False)

        # Plot ROC curve
        if not np.isnan(roc):
            fpr, tpr, _ = roc_curve(y_true, y_proba)
            plt.figure(figsize=(6,4))
            plt.plot(fpr, tpr, label=f"{prefix} (AUC = {roc:.3f})")
            plt.plot([0,1],[0,1],"--")
            plt.xlabel("False Positive Rate")
            plt.ylabel("True Positive Rate")
            plt.title(f"ROC Curve - {prefix}")
            plt.legend()
            plt.tight_layout()
            plt.savefig(os.path.join(OUTPUT_DIR, f"roc_{prefix}.png"))
            plt.close()

        # Confusion matrix heatmap save
        plt.figure(figsize=(4,3))
        sns.heatmap(cm, annot=True, fmt="d", cmap="Blues")
        plt.title(f"Confusion Matrix - {prefix}")
        plt.xlabel("Predicted")
        plt.ylabel("Actual")
        plt.tight_layout()
        plt.savefig(os.path.join(OUTPUT_DIR, f"cm_{prefix}.png"))
        plt.close()

    # Evaluate logistic
    eval_and_save(y_test, y_pred_log, y_proba_log, "Logistic")

    # Evaluate RF
    eval_and_save(y_test, y_pred_rf, y_proba_rf, "RandomForest")

    # Evaluate XGB if available
    if XGBOOST_AVAILABLE:
        eval_and_save(y_test, y_pred_xgb, y_proba_xgb, "XGBoost")

    # Save best model (choose RF if gridsearch done)
    model_to_save = best_rf if 'best_rf' in locals() else pipe_log
    model_path = os.path.join(OUTPUT_DIR, "model_turnover.joblib")
    joblib.dump(model_to_save, model_path)
    print(f"Saved model to: {model_path}")

    # Save sample predictions
    sample_out = X_test.copy()
    sample_out["y_true"] = y_test.values
    sample_out["rf_pred"] = y_pred_rf
    sample_out["rf_proba"] = y_proba_rf
    sample_out.to_csv(os.path.join(OUTPUT_DIR, "predictions_sample.csv"), index=False)
    print("Saved sample predictions to predictions_sample.csv")

if __name__ == "__main__":
    main()
